File: solutions/typescript/robot-simulator/robot-simulator.ts
export class InvalidInputError extends Error {
  constructor(message?: string) {
    super();
    this.message = message || 'Invalid Input';
  }
}

type Direction = 'north' | 'east' | 'south' | 'west'
type Coordinates = [number, number]

const directions: Direction[] = ['north', 'east', 'south', 'west'];

function getOffset(direction: Direction): Coordinates {
  switch (direction) {
    case 'north':
      return [0, 1];
    case 'east':
      return [1, 0];
    case 'south':
      return [0, -1];
    case 'west':
      return [-1, 0];
  }
}

export class Robot {
  bearing: Direction;
  coordinates: Coordinates;

  constructor() {
    this.bearing = 'north';
    this.coordinates = [0, 0];
  }

  place({x, y, direction}: { x: number; y: number; direction: string }) {
    if (direction !== 'north'
     && direction !== 'east'
     && direction !== 'south'
     && direction !== 'west') {
      throw new InvalidInputError();
    }
    this.bearing = direction;
    this.coordinates = [x, y];
  }

  evaluate(instructions: string) {
    instructions.split('').forEach((c) => {
      switch (c) {
        case 'A': {
          const [offsetX, offsetY] = getOffset(this.bearing);
          this.coordinates[0] += offsetX;
          this.coordinates[1] += offsetY;
          break;
        }
        case 'L':
          this.offsetBearing(-1);
          break;
        case 'R':
          this.offsetBearing(1);
          break;
        default:
          throw new InvalidInputError();
      }
    })
  }

  private offsetBearing(n: number) {
    const l = directions.length;
    const i = (l + directions.indexOf(this.bearing) + n) % l;
    this.bearing = directions[i];
  }
}

File: solutions/python/robot-simulator/robot_simulator.py
from enum import Enum


class Direction(Enum):
    EAST = 0
    NORTH = 1
    WEST = 2
    SOUTH = 3


EAST = Direction.EAST
NORTH = Direction.NORTH
WEST = Direction.WEST
SOUTH = Direction.SOUTH

type Coordinates = tuple[int, int]


def direction_offset(direction: Direction) -> Coordinates:
    match direction:
        case Direction.EAST:
            return 1, 0
        case Direction.WEST:
            return -1, 0
        case Direction.NORTH:
            return 0, 1
        case Direction.SOUTH:
            return 0, -1


class Robot:
    direction: Direction
    coordinates: Coordinates

    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def move(self, param: str):
        for char in param:
            match char:
                case 'L':
                    self.offset_direction(+1)
                case 'R':
                    self.offset_direction(-1)
                case 'A':
                    self.advance(direction_offset(self.direction))

    def offset_direction(self, n: int):
        v = self.direction.value
        self.direction = Direction((4 + v + n) % 4)

    def advance(self, coords: Coordinates):
        (x, y) = self.coordinates
        (off_x, off_y) = coords
        self.coordinates = (x + off_x, y + off_y)
